libbpf-rs/src/iter.rs

```diff
@@ -0,0 +1,41 @@
+use nix::{libc, unistd};
+use std::io;
+
+use crate::*;
+
+/// Represents a bpf iterator for reading kernel data structures. This requires
+/// Linux 5.8.
+///
+/// This implements [`std::io::Read`] for reading bytes from the iterator.
+/// Methods require working with raw bytes. You may find libraries such as
+/// [`plain`](https://crates.io/crates/plain) helpful.
+pub struct Iter {
+    fd: i32,
+}
+
+impl Iter {
+    pub fn new(link: &Link) -> Result<Self> {
+        let link_fd = link.get_fd();
+        let fd = unsafe { libbpf_sys::bpf_iter_create(link_fd) };
+        if fd < 0 {
+            return Err(Error::System(-fd));
```

Comment:
you use last_os_error() in read, shouldn't that be used here as well?

Reply:
We could do that but we'd need to return `std::result::Result<usize, std::io::Error>` rather than `std::result::Result<usize, crate::Error>` and I don't think that's any better? Ideally, everything in this library should be using `crate::Error`. We don't in read since we have to `impl io::Read`.

Reply:
as it is right now, -fd will be -1 which is -EPERM, which is a lie.

Reply:
oh right, my bad. Should be fixed now.